### PR TITLE
feat(notifications): adds providers to notification settings endpint

### PR DESCRIPTION
**Pull request by scefali**
This PR adds a new `v2` option to the `UserNotificationSettingsDetailsEndpoint` which has the following fields in the response:
* `preferences`: This is the new name for settings. This reflects the _preferences_ for a notification with a provider but it's no guarantee that a notification can be sent on that channel. This is the full response of the v1 version of this endpoint.
* `providers`: These are the providers that the user has available based on the rows in the `Identity` model. Note that since in the UI sets the provider globally, we can provide a global set of providers for that user.

The intention behind this new API is so that in the UI settings, we only show options for the providers the user has available. For example, if they don't have their Slack identity mapped then we won't get `slack` in the list of providers and we won't render it as an option.

The migration path is to first land this new backend change, then add `v2=1` to the UI and make the modifications above, then remove the old format. Since this API has no usage by customers, we can make this breaking API change without issue once the UI is updated.

### CODE REVIEW

**Comment by ceorourke** [comment]
@AniketDas-Tekky can confirm but I don't think we consider adding something to an API response to be a breaking change, only removing a field or changing a field's response, so you don't need to bother w/ the v2 stuff here.

File: src/sentry/notifications/helpers.py [hunk 1]
```diff
@@ -583,3 +589,29 @@ def get_values_by_provider(
         get_value_for_actor(notification_settings_by_scope, recipient),
         get_value_for_parent(notification_settings_by_scope, parent_id, type),
     )
+
+
+def get_providers_for_recipient(recipient: User | Team) -> Iterable[ExternalProviders]:
```
Comment (ceorourke) [inline src/sentry/notifications/helpers.py:594]: Would you be able to use one of the existing methods like `where_should_recipient_be_notified`? I guess that one takes a single type ... it seems weird that we don't already have this though. Or maybe `transform_to_notification_settings_by_recipient`?

**Comment by scefali** [comment]
@ceorourke The response has changed in a breaking way. Previously the response was like this:
```
{
  "workflow": {
    "user": {
      "me": {
        "email": "always",
        "slack": "always"
      }
    }
  }
}
```

Now it's like:
```
{
  "preferences": {
    "workflow": {
      "user": {
        "me": {
          "email": "never",
          "slack": "never"
        }
      }
    }
  },
  "providers": [
    "slack",
    "email"
  ]
}
```
It has the same info, but it's more nested to support additional data.

File: src/sentry/notifications/helpers.py (hunk 1 above)
Comment (scefali) [reply src/sentry/notifications/helpers.py:594]: @ceorourke No, I cannot use that for the fundamental reason that `where_should_recipient_be_notified` doesn't check the `Identity` table. What might make more sense, in the long run, is for `where_should_recipient_be_notified` to call `get_providers_for_recipient` and use that to determine how we should notify the user.

**Comment by ceorourke** [comment]
Re: the response (quoting it seems excessive) I guess I imagined it'd just be adding the providers list to the response, is there a reason that wouldn't work?

**Comment by scefali** [comment]
> Re: the response (quoting it seems excessive) I guess I imagined it'd just be adding the providers list to the response, is there a reason that wouldn't work?

@ceorourke Yes. Previously, to get the workflow settings you do something like `response.data["workflow"]`. Now you need to do `response.data["preferences"]["workflow"]`. So any code that relied on the shape of the response (such as the UI) would break with the change.

**Comment by markstory** [comment]
> Now you need to do response.data["preferences"]["workflow"]. So any code that relied on the shape of the response (such as the UI) would break with the change.

Do any of the notification names overlap with `providers`? If not, do you need the extra layer of attributes? Skipping that would make the API changes backwards compatible.

File: src/sentry/api/endpoints/user_notification_settings_details.py [hunk 2]
```diff
@@ -29,16 +30,25 @@ def get(self, request: Request, user: User) -> Response:
         """
 
         type_option = validate_type_option(request.GET.get("type"))
+        v2_format = request.GET.get("v2") == "1"
```
Comment (AniketDas-Tekky) [inline src/sentry/api/endpoints/user_notification_settings_details.py:33]: can we do a `version=2.0` instead. Incase we make more breaking changes in the future, we don't have to keep adding arguments.

**Comment by scefali** [comment]
@markstory @AniketDas-Tekky To be clear, I am not suggesting that we keep two different response formats in the long term. IMO it makes more sense to do a breaking change once I update the front-end to use the new response format. This endpoint has no use as a public API so there is no reason to maintain two formats/version. This `v2` thing I'm adding is only going to last until I made the front-end change. Once that's done, I'll remove the logic related to it and we'll always return a new format. If this API was something customers are using other than through the UI then it would be a different story.

**Comment by AniketDas-Tekky** [comment]
@scefali Just confirmed that no external requests are being made to this endpoint. LGTM

**Comment by scefali** [comment]
> Do any of the notification names overlap with `providers`? If not, do you need the extra layer of attributes? Skipping that would make the API changes backwards compatible.

@markstory  I'm opting for the approach to have the cleanest final API because I can make breaking changes here once I update the UI

**Review by AniketDas-Tekky — APPROVED** [review]

**Pull request merged**